File: test-fixtures/languages/go/micro/auth_service.go
package micro

import (
	"database/sql"
	"os/exec"
)

func VerifyToken(db *sql.DB, token string) (string, error) {
	query := "SELECT user_id FROM tokens WHERE token = '" + token + "'"
	row := db.QueryRow(query) // sink: SQL injection
	var userID string
	err := row.Scan(&userID)
	if err != nil {
		return "", err
	}
	return userID, nil
}

func RunAdminCommand(userID string, cmd string) error {
	if userID != "" {
		return exec.Command("notify-admin", cmd).Run() // sink: command injection
	}
	return nil
}

File: test-fixtures/languages/rust/micro/auth_service.rs
use rusqlite::Connection;
use std::process::Command;

pub fn verify_token(conn: &Connection, token: &str) -> Option<String> {
    let query = format!("SELECT user_id FROM tokens WHERE token = '{}'", token);
    let mut stmt = conn.prepare(&query).ok()?; // sink: SQL injection
    let user_id: String = stmt.query_row([], |row| row.get(0)).ok()?;
    Some(user_id)
}

pub fn run_admin_command(user_id: &str, cmd: &str) {
    if !user_id.is_empty() {
        let full_cmd = format!("notify-admin {}", cmd);
        Command::new("sh")
            .arg("-c")
            .arg(&full_cmd)
            .output()
            .expect("failed"); // sink: command injection
    }
}
